handler/src/test/java/io/netty/handler/stream/ChunkedWriteHandlerTest.java

```diff
@@ -227,6 +228,207 @@ public long progress() {
+    @Test
+    public void testFailureWhenLastChunkFailed() throws IOException {
+        ChannelOutboundHandlerAdapter failLast = new ChannelOutboundHandlerAdapter() {
+            private int passedWrites;
+
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+                if (++this.passedWrites < 4) {
+                    ctx.write(msg, promise);
+                } else {
+                    promise.tryFailure(new RuntimeException());
+                }
+            }
+        };
+
+        EmbeddedChannel ch = new EmbeddedChannel(failLast, new ChunkedWriteHandler());
+        ChannelFuture r = ch.writeAndFlush(new ChunkedFile(TMP, 1024 * 16)); // 4 chunks
+        ch.close().awaitUninterruptibly();
+
+        assertFalse(r.isSuccess());
+        assertTrue(r.cause() instanceof RuntimeException);
+
+        // 3 out of 4 chunks were already written
+        int read = 0;
+        for (;;) {
+            ByteBuf buffer = ch.readOutbound();
+            if (buffer == null) {
+                break;
+            }
+            read += buffer.readableBytes();
+            buffer.release();
+        }
+
+        assertEquals(1024 * 16 * 3, read);
+    }
+
+    @Test
+    public void testDiscardPendingWritesOnInactive() throws IOException {
+
+        final AtomicBoolean closeWasCalled = new AtomicBoolean(false);
+
+        ChunkedInput<ByteBuf> notifiableInput = new ChunkedInput<ByteBuf>() {
+            private boolean done;
+            private final ByteBuf buffer = Unpooled.copiedBuffer("Test", CharsetUtil.ISO_8859_1);
+
+            @Override
+            public boolean isEndOfInput() throws Exception {
+                return done;
+            }
+
+            @Override
+            public void close() throws Exception {
+                buffer.release();
+                closeWasCalled.set(true);
+            }
+
+            @Deprecated
+            @Override
+            public ByteBuf readChunk(ChannelHandlerContext ctx) throws Exception {
+                return readChunk(ctx.alloc());
+            }
+
+            @Override
+            public ByteBuf readChunk(ByteBufAllocator allocator) throws Exception {
+                if (done) {
+                    return null;
+                }
+                done = true;
+                return buffer.retainedDuplicate();
+            }
+
+            @Override
+            public long length() {
+                return -1;
+            }
+
+            @Override
+            public long progress() {
+                return 1;
+            }
+        };
+
+        EmbeddedChannel ch = new EmbeddedChannel(new ChunkedWriteHandler());
+
+        // write 3 messages and close channel before flushing
+        ChannelFuture r1 = ch.write(new ChunkedFile(TMP));
+        ChannelFuture r2 = ch.write(new ChunkedNioFile(TMP));
+        ch.write(notifiableInput);
+        ch.close().awaitUninterruptibly();
```

Comment:
just use `ch.finish()` and assert the return value.

Reply:
@normanmaurer `ch.finish` would return `false` because we effectively don't do any writes here. All messages are still in `ChunkedWriteHandler` queue, until `flush`. Meaning both `inboundMessages` and `outboundMessages` queues are empty.

Reply:
@kachayev sure which is fine just use `assertFalse(...)` as this is what you expect, or what I am missing ?

Reply:
@normanmaurer Oh, yeah. You're right! 🤦‍♂️It's just mentally hard to read `assertFalse(ch.finish())` as "all good, just no messages" :) I'll leave a comment why this is expected.

Reply:
@normanmaurer Update once again. I also "fixed" a couple of older tests to be consistent in terms of `ch.finish()` usage.